src/Pate/Verification/PairGraph.hs

```diff
@@ -226,6 +263,98 @@ getReturnVectors ::
 getReturnVectors gr fPair =
   fromMaybe mempty (Map.lookup fPair (pairGraphReturnVectors gr))
 
+-- | Look up the current abstract domain for the given graph node.
+getCurrentDomain ::
+  PairGraph sym arch ->
+  GraphNode arch ->
+  Maybe (AbstractDomain sym arch)
+getCurrentDomain pg nd = Map.lookup nd (pairGraphDomains pg)
+
+-- | If an observable counterexample has not already be found
+--   for this block pair, run the given action to check if there
+--   currently is one.
+considerObservableEvent :: Monad m =>
+  PairGraph sym arch ->
+  PPa.BlockPair arch ->
+  (m (Maybe (ObservableCounterexample sym (MM.ArchAddrWidth arch)), PairGraph sym arch)) ->
+  m (PairGraph sym arch)
+considerObservableEvent gr bPair action =
+  case Map.lookup bPair (pairGraphObservableReports gr) of
+    -- we have already found observable event differences at this location, so skip the check
+    Just _ -> return gr
+    Nothing ->
+      do (mcex, gr') <- action
+         case mcex of
+           Nothing  -> return gr'
+           Just cex -> return gr'{ pairGraphObservableReports = Map.insert bPair cex (pairGraphObservableReports gr) }
+
+-- | If a program desync has not already be found
+--   for this block pair, run the given action to check if there
+--   currently is one.
+considerDesyncEvent :: Monad m =>
+  PairGraph sym arch ->
+  PPa.BlockPair arch ->
+  (m (Maybe (TotalityCounterexample (MM.ArchAddrWidth arch)), PairGraph sym arch)) ->
+  m (PairGraph sym arch)
+considerDesyncEvent gr bPair action =
+  case Map.lookup bPair (pairGraphDesyncReports gr) of
+    -- we have already found observable event differences at this location, so skip the check
+    Just _ -> return gr
+    Nothing ->
+      do (mcex, gr') <- action
+         case mcex of
+           Nothing  -> return gr'
+           Just cex -> return gr'{ pairGraphDesyncReports = Map.insert bPair cex (pairGraphDesyncReports gr) }
```

Comment:
Are these two basically just the same, but with different accessors? Should we factor out this structure with e.g., lenses to reduce duplication? Probably not a huge issue, since they are pretty simple

Reply:
They are basically the same. I'm not convinced it's worth abstracting using lenses for just these two... but we certainly could.